src/app/digital/public/api/impl/DigitalComponent.ts

```diff
@@ -16,4 +19,38 @@ export class DigitalComponentImpl extends ComponentImpl<
     public override get info(): DigitalComponentInfo {
         return new DigitalComponentInfoImpl(this.circuit, this.kind);
     }
+
+    public override firstAvailable(portGroup: string): DigitalPort | undefined {
+        const Ports = this.internal.getPortsForComponent(this.id);
+
+        if (portGroup === "input") {
+            const portTypes = this.info.inputPortGroups;
+            if (!portTypes.includes('inputs')) {
+                return undefined;
+            } else {
+                for(var p of Ports.values()) {
+                    let portObject = this.internal.getPortByID(p);
+                    let portWire = this.internal.getWiresForPort(p); // no wires = available
+                    if (portObject !== undefined && portObject.group === 'inputs' && portWire.size === 0) {
+                        return new DigitalPortImpl(this.circuit, p);  
+                    }
+                }
+            }
+        }
+
+        if (portGroup === "output") {
+            const portTypes = this.info.outputPortGroups;
+            if (!portTypes.includes('outputs')) {
+                return undefined;
+            } else {
+                for(var p of Ports.values()) {
+                    let portObject = this.internal.getPortByID(p);
+                    let portWire = this.internal.getWiresForPort(p); // no wires = available
+                    if (portObject !== undefined && portObject.group === 'outputs') {
+                        return new DigitalPortImpl(this.circuit, p);  
+                    }
+                }
+            }
```

Comment:
These code blocks are very similar, try and see if you can condense them

Also note that you can make use of `.find` to avoid writing a for loop explicitly